helper/schema/schema.go

```diff
@@ -145,7 +145,12 @@ type Schema struct {
 	// The following fields are only valid for a TypeSet type.
 	//
 	// Set defines a function to determine the unique ID of an item so that
-	// a proper set can be built.
```

Comment:
Please correct if these comments are wrong, but they seem necessary considering https://github.com/hashicorp/terraform/issues/10520, https://github.com/hashicorp/terraform/pull/2336 and https://groups.google.com/forum/#!topic/terraform-tool/tsnWzWfkzaQ.  

Took a long time for me to understand what was going on.

Which brings up another issue.  We've got a lot of properties in azurerm (e.g. VMs/VMSS) that get ignored because they're in sets that define a set function that includes only certain elements.  I don't see any way to fix this, because that would modify the hash and cause everybody's resources to be updated.

Part of the problem is the documentation here, which to me implies that an "ID" is *not* a hash, but rather a way for hashicorp to keep track of a particular object in a set with its cloud counterpart, and thus be able to detect when non-identity properties are changed, and make them to the appropriate cloud resource.

Instead, it appears that the set function is solely a hash code, which has quite a few disadvantages:
1) Any change in the properties acts as if the entire set has changed (and if any of the properties in the set are ForceNew=true, then the entire resource gets re-created)
2) The diff shown to the user by this is way more complex than it needs to be and is being confusing
3) There's not a way to modify just certain properties of that object
4) There's a lot of code at least in azurerm that has incorrect set functions and thus properties that get ignored, with no clear way to fix these bugs

Seems like this design may need review.

The design mentioned in